themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,196 @@ function toggleVisibility(e) {
+	const p = document.getElementsByTagName('p');
+	let scrolled = false;
+	for (let searchTarget of [...h2, ...p]) {
+		const a = searchTarget.querySelector('a');
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		if (highlightText(searchTarget, search) && !scrolled) {
+			scrolled = true;
+			searchTarget.scrollIntoView();
+		}
+	}
+}
+
+function menuItemsSearch(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const titles = nav.querySelectorAll('.title');
+	for (let title of titles) {
+		const menuItem = title.parentElement;
+		for (let tag of getTags(menuItem)) {
+			if (search && tag.toLowerCase().indexOf(search) >= 0) {
+				appendTagText(menuItem, tag, search);
+				const dataShowTag = menuItem.querySelector('.tag');
+				highlightText(dataShowTag, search);
+				break;
+			}
+			removeTagText(menuItem, search);
+		}
+		highlightText(title, search);
+	}
+}
+
+function appendTagText(menuItem, tag, search) {
+	menuItem.classList.remove('hidden');
+	if (!menuItem.querySelector('.tag')) {
+		menuItem.insertBefore(document.createElement('a'), menuItem.querySelector('.title'));
+		menuItem.querySelector('a').classList.add('tag');
+	}
+	const title = menuItem.querySelector('.title');
+	title.classList.add('menu-item');
+	if (!menuItem.querySelector('.border')) {
+		const border = menuItem.insertBefore(document.createElement('span'), menuItem.querySelector('.title'));
+		border.classList.add('border');
+		border.textContent = '/ ';
+	}
+	const menuTag = menuItem.querySelector('.tag');
+	menuTag.classList.remove('hidden');
+	menuTag.textContent = formatTag(tag, search);
+	menuTag.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+}
+
+function removeTagText(menuItem, search) {
+	const title = menuItem.querySelector('.title');
+	if (search) {
+		menuItem.classList.add('hidden');
+	}
+	else {
+		menuItem.classList.remove('hidden');
+	}
+	const tag = menuItem.querySelector('.tag');
+	if (tag) {
+		tag.classList.add('hidden');
+	}
+	title.classList.remove('menu-item');
+	if (menuItem.querySelector('.border')) {
+		menuItem.removeChild(menuItem.querySelector('.border'));
+	}
+}
+
+function highlightText(item, search) {
+	const { textContent } = item;
+	const searchContains = new RegExp(search, 'ig');
+	if (search && searchContains.test(textContent)) {
+		item.innerHTML = textContent.replace(searchContains, elem => `<span class="highlight">${elem}</span>`);
+		item.parentElement.classList.remove('hidden');
+		return true;
+	}
+	item.innerHTML = textContent;
+}
+
+function formatTag(tag, search) {
+	let tagsContent = split(tag);
+	const index = getIndexOfTag(tagsContent, search);
+	const visibleWordCount = 2;
```

Comment:
please move this constants to the top of file and use upper case

VISIBLE_MENU_TAG_COUNT = 2;